### DISCUSSION TITLE
Enable Vercel Toolbar/Preview Comments on Production Storybook

### DISCUSSION THREAD

**Discussion by kfayelun** [storybookjs/storybook#26092] [Help]
### Summary

I'm having trouble enabling Vercel Toolbar on a production build of Storybook, and would greatly appreciate some help 😊 We want to enable this so our design system users easily can give us feedback on the documentation. We use it a lot for other apps, and it's a great way to get quick feedback, straight from users. Can recommend! The Vercel documentation says to do [this](https://vercel.com/docs/workflow-collaboration/comments/in-production-and-localhost), but I'm having a hard time translating that to the Storybook setup. 
I've tried a couple approaches, but all using `manager-head.html` or `preview-head.html` in combination with the corresponding field in the `main.ts` config; `managerHead` or `previewHead`. Using `managerHead`, I can see the toolbar show up in the markup, wrapping the application, and itself wrapped by `<body>`, and I assume that is the correct way to go (?). Here's what that looks like: 

In `apps/docs/.storybook/main.ts`:
```
managerHead: (head) => `
    ${head}
    <VercelToolbar />
  `,
```
  
In `apps/docs/.storybook/manager-head.html`:
```
<script type="module">
  import { VercelToolbar } from "@vercel/toolbar";
</script>
```
Very unsure about this 👆 part. Guessing this is where it goes wrong.
But how do I get the app to import this when the `managerHead` just reads a string? 
Do I have to use a `require()` somewhere? Ideally I would like to enable/disable the functionality with an env var, but that is 2nd priority to just getting to work.

In `apps/docs/package.json`:
```
"dependencies": {
"@vercel/toolbar": "^0.1.11",
}
```

As I would like to test this before pushing to production 😬 I used `vercel build --prod` to build a production build, and `serve storybook-static` to run that build. That yields no status bar and this error message in the console:
```Uncaught TypeError: Failed to resolve module specifier "@vercel/toolbar". Relative references must start with either "/", "./", or "../". ```

### Additional information

We're using turborepo `^1.10.0`, and the storybook app is under `apps/docs` as turborepos docs suggest. 
I've removed confidential information and replaced it with `***` where it makes sense. 

```
{
  "name": "***",
  "private": true,
  "packageManager": "npm@9.9.2",
  "repository": {
    "type": "git",
    "url": "***"
  },
  "scripts": {
    "storybook": "storybook dev -p 6006",
    "build:storybook": "storybook build",
    "preview:storybook": "serve storybook-static",
    "dev": "npm run storybook",
    "clean": "rm -rf .turbo && rm -rf node_modules && rm -rf storybook-static",
    "chromatic": "npx chromatic",
    "type-check": "tsc --noEmit",
    "lint": "eslint . --ext js,ts,tsx,md,mdx",
    "lint:fix": "npm run lint -- --fix"
  },
  "dependencies": {
    "@stripe/react-stripe-js": "^2.4.0",
    "@stripe/stripe-js": "^2.2.2",
    "@vercel/toolbar": "^0.1.11",
    "axios": "^1.6.3",
    "react": "^18.2.0",
    "react-dom": "^18.2.0"
  },
  "devDependencies": {
    "@phosphor-icons/react": "^2.0.10",
    "@storybook/addon-essentials": "^7.4.6",
    "@storybook/addon-interactions": "^7.4.6",
    "@storybook/addon-themes": "^7.4.6",
    "@storybook/blocks": "^7.4.6",
    "@storybook/manager-api": "^7.4.6",
    "@storybook/react": "^7.4.6",
    "@storybook/react-vite": "^7.4.6",
    "@storybook/testing-library": "^0.2.0",
    "@types/react": "^18.2.24",
    "@types/react-dom": "^18.2.6",
    "@vitejs/plugin-react": "^4.2.1",
    "autoprefixer": "^10.4.14",
    "chromatic": "^6.19.9",
    "eslint-config-foo": "*",
    "eslint-config-custom": "*",
    "postcss": "^8.4.31",
    "serve": "^14.2.0",
    "storybook": "^7.4.6",
    "tailwindcss": "^3.3.3",
    "typescript": "^5.2.2",
    "vite": "^4.5.1",
    "vite-plugin-turbosnap": "^1.0.2"
  }
}
```

`main.ts`
```
import type { StorybookConfig } from "@storybook/react-vite";
import { mergeConfig } from "vite";
import turbosnap from "vite-plugin-turbosnap";
import * as path from "path";

/** As the Storybook typescript option types are conditional we need some infer magic to get the type sorted */
type StorybookConfigTypescript = StorybookConfig extends {
  typescript?: infer E;
}
  ? E
  : never;
type StorybookConfigTypescriptReactDocgen = StorybookConfigTypescript extends {
  reactDocgenTypescriptOptions?: infer E;
}
  ? E
  : never;

const typePropFilter: StorybookConfigTypescriptReactDocgen["propFilter"] = (
  prop,
) => (prop.parent ? !/node_modules/.test(prop.parent.fileName) : true);

const config: StorybookConfig = {
  stories: [
    {
      directory: "../src/",
      files: "readme.mdx",
    },
    {
      directory: "../src/docs/",
      titlePrefix: "Guidelines",
      files: "@(theming|customizing).mdx",
    },
    {
      directory: "../src/experimental",
      titlePrefix: "Experimental",
    },
  ],
  addons: [
    getAbsolutePath("@storybook/addon-essentials"),
    getAbsolutePath("@storybook/addon-interactions"),
    getAbsolutePath("@storybook/addon-themes"),
  ],
  staticDirs: ["./public"],
  framework: {
    // The below seems like a bad type on the SB side so just gonna ignore it
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    name: getAbsolutePath("@storybook/react-vite"),
    options: {},
  },
  core: {
    // The below seems like a bad type on the SB side so just gonna ignore it
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    builder: getAbsolutePath("@storybook/builder-vite"),
  },
  managerHead: (head) => `
    ${head}
    <VercelToolbar />
  `,
  async viteFinal(config, { configType }) {
    return mergeConfig(config, {
      plugins:
        configType === "PRODUCTION"
          ? [turbosnap({ rootDir: config.root ?? process.cwd() })]
          : [],
      resolve: {
        alias:
          // see README for this
          configType === "DEVELOPMENT"
            ? {
                "***": path.resolve(
                  __dirname,
                  "../../../packages/components/src",
                ),
// tried the below to see if I could get it to work in development, did not work
                "@vercel/toolbar": path.resolve(
                  __dirname,
                  "../node_modules/.cache/sb-vite/@vercel/toolbar.js",
                ),
              }
            : {
                "***": path.resolve(
                  __dirname,
                  "../../../node_modules/***",
                ),
              },
      },
      // ...And any other config you need to change...
    });
  },
  typescript: {
    check: false, // only for webpack
    reactDocgen: "react-docgen-typescript",
    reactDocgenTypescriptOptions: {
      shouldExtractLiteralValuesFromEnum: true,
      propFilter: typePropFilter,
    },
  },
  features: {
    storyStoreV7: true,
  },
  docs: {
    autodocs: true,
  },
};
module.exports = config;

function getAbsolutePath(value: string): string {
  return path.dirname(require.resolve(path.join(value, "package.json")));
}
```

How the toolbar now shows up in the markup (seemed excessive to add the markup as text, but I could do that too if its better. I did describe this behaviour in the text above though.): 
![Screenshot 2024-02-19 at 10 59 09](https://github.com/storybookjs/storybook/assets/4848101/3ebdd80c-94da-4d25-b49d-7bdc8a7b3fb9)

**Comment by gdborton** [comment]
@kfayelun 👋 EM for the team that owns the toolbar. This would be easiest to debug if you have a repro.  I've used storybook in the past, but I'm not intimately familiar with it.

```
<script type="module">
  import { VercelToolbar } from "@vercel/toolbar";
</script>
```

Looks like you're trying to import a module in html, which will fail as your document isn't aware of where `@vercel/toolbar` lives... you _might_ have more luck trying `https://unpkg.com/@vercel/toolbar` here.

I'm not sure that'll work either though... if not you can try adding the script tag directly (this is what the toolbar package does):
```html
<script src="https://vercel.live/_next-live/feedback/feedback.js" id="vercel-toolbar" data-explicit-opt-in="true"></script>
```

**Reply by kfayelun** [reply]
Hey @gdborton ! 

Thanks so much for your quick reply. I tried your second option and just pushed it to production (internal tool so kinda ok), as I thought that would work. But I get ERR_BLOCKED_BY_CLIENT in the console and no Vercel Toolbar. 
<img width="1169" alt="Screenshot 2024-03-01 at 17 51 52" src="https://github.com/storybookjs/storybook/assets/4848101/0311c085-5fdc-42c9-aedc-dd87adf734b4">

I see that the script tag is in the `<head>` element on the page, same way it is on the other "normal" web apps we have which has the toolbar, so I guess that is correct. 
![vercelToolbarMarkup](https://github.com/storybookjs/storybook/assets/4848101/cf920164-3408-4c92-97fa-aea3e2377ec4)
It's missing the font download though, which I guess I also need? This: `<link href="https://vercel.live/fonts" rel="stylesheet">`

How does [this way](https://vercel.com/docs/workflow-collaboration/comments/in-production-and-localhost) of installing the toolbar get around cross-site requests (if that is the issue here)? 
I first guessed that for preview builds, as those are also hosted on vercel, there wouldn't be an issue. But I see that its the same error and no toolbar for the `foo.vercel-preview.company.domain` as well. So that's not why.

I assume I will have to figure this out no matter which way I want to install the toolbar?  

I checked that I have enabled Preview Comments for this project in Vercel. As far as I can tell theres no CORS policy or similar inside storybook (?), and Chromatic isn't relevant here as we are hosting it on Vercel. So not sure how to proceed on this now.

**Reply by gdborton** [reply]
@kfayelun sorry I'm swamped w/ notifications atm and missed your reply.

We've got details for how to handle CORS issues here:
https://vercel.com/docs/workflow-collaboration/comments/specialized-usage#using-a-content-security-policy

Although it looks like the failed requests that you're seeing _may_ be from an ad-blocker.

It does seem like the script is running (those event/tick calls are coming from the feedback script).

I'd expect to see the `<vercel-live-feedback />` custom element loaded here.
